### PR TITLE
Escape the contents of metadata strings

### PR DESCRIPTION
**Pull request by whitequark**
Also, allow all printable ASCII characters in IR string literals,
including metadata strings and character array initializers.

### CODE REVIEW

File: llvmlite/ir/values.py [hunk 1]
```diff
@@ -14,13 +14,15 @@
 
 _VALID_CHARS = (frozenset(map(ord, string.ascii_letters)) |
                 frozenset(map(ord, string.digits)) |
-                frozenset('._-$'))
+                frozenset(map(ord, ' !#$%&\'()*+,-./:;<=>?@[]^_`{|}~')))
```
Comment (whitequark) [inline llvmlite/ir/values.py]: This was actually broken before, since it didn't have `map(ord, ...)`.

**Comment by whitequark** [comment]
I strongly suspect based on the nature of the fix I had to do, but haven't bothered to prove, that `_escape_string` has never properly worked on Python 2.

**Comment by whitequark** [comment]
Yeah, it would have broke if someone passed it a `bytes` constant. Have you yet considered doing the appropriate actions to handle Python 2, i.e. killing it with fire?

**Comment by whitequark** [comment]
The Appveyor error is bogus.

File: llvmlite/ir/values.py (hunk 1 above)
Comment (pitrou) [inline llvmlite/ir/values.py]: How about the reverse, i.e. only use `ord` on Python 3?

File: llvmlite/ir/values.py [hunk 2]
```diff
@@ -14,13 +15,15 @@
...
 
 
 def _escape_string(text, _map={}):
     """
     Escape the given bytestring for safe use as a LLVM array constant.
     """
+    if isinstance(text, six.text_type):
```
Comment (pitrou) [inline llvmlite/ir/values.py]: I'm not sure why this is necessary? You should never pass a unicode string here...

**Comment by pitrou** [comment]
> Have you yet considered doing the appropriate actions to handle Python 2, i.e. killing it with fire?

I would like that, but many of our users (including professional customers of Continuum products) still use Python 2...

File: llvmlite/ir/values.py (hunk 2 above)
Comment (whitequark) [inline llvmlite/ir/values.py]: Initially I did `isinstance(text, str)` since otherwise the assert below would fail on Python 3, but that failed on Python 2 because of the superfluous `encode` call...

Comment (pitrou) [inline llvmlite/ir/values.py]: But why did you need that at all? You should never pass a text string there, only a bytes or bytearray object.

Comment (pitrou) [inline llvmlite/ir/values.py]: In other words, if you need an encode() call, it should be in `metadata._get_reference()`, not here.

Comment (whitequark) [inline llvmlite/ir/values.py]: I don't see why. This code is common for all double-quoted strings in the IR, whether originally strings (and thus user-facing UTF-8 identifiers) or byte arrays (probably emitted by the compiler).

Comment (pitrou) [inline llvmlite/ir/values.py]: Passing an unicode string as an array constant feels like a programmer error to me, so I'd rather raise an exception. I don't know about metadata strings.

File: llvmlite/ir/values.py (hunk 1 above)
Comment (whitequark) [inline llvmlite/ir/values.py]: Well, that adds a py2/py3 branching, which means more opportunity for bugs. I don't see this as necessary since this code is executed exactly once.

Comment (whitequark) [inline llvmlite/ir/values.py]: Well, I guess there's this branch below...
```

    if six.PY2 and isinstance(text, bytes):
        buf = [_map[ord(ch)] for ch in text]
    else:
        buf = [_map[ch] for ch in text]
```
so using `ord()` would pessimize py3. This is a rather hot function since IR contains many strings, and I don't think that should be done.

File: llvmlite/ir/values.py (hunk 2 above)
Comment (whitequark) [inline llvmlite/ir/values.py]: Removed the `encode` call.

**Comment by whitequark** [comment]
Updated.

File: llvmlite/ir/instructions.py [hunk 3]
```diff
@@ -162,7 +162,7 @@ def descr(self, buf):
 class PredictableInstr(Instruction):
 
     def set_weights(self, weights):
-        operands = [MetaDataString(self.module, "branch_weights")]
+        operands = [MetaDataString(self.module, b"branch_weights")]
```
Comment (pitrou) [inline llvmlite/ir/instructions.py]: For the required, is this change required by the rest of the patch?

File: llvmlite/ir/values.py (hunk 1 above)
Comment (pitrou) [inline llvmlite/ir/values.py]: Hmm, I understand. Perhaps `_map` can contain both cases on py2 instead? That way, the branch would happen at construction time. Something like:
```
if six.PY2:
    _map[chr(ch)] = _map[ch]
```

File: llvmlite/ir/instructions.py (hunk 3 above)
Comment (whitequark) [inline llvmlite/ir/instructions.py]: Yes, since you've requested the encode call to be removed. This worked before by accident.

File: llvmlite/tests/test_ir.py [hunk 4]
```diff
@@ -202,12 +202,18 @@ def test_metadata_2(self):
         mod = self.module()
         mod.add_metadata([ir.Constant(ir.IntType(32), 123)])
         mod.add_metadata([ir.Constant(ir.IntType(32), 321),
-                          ir.MetaDataString(mod, "kernel")])
+                          ir.MetaDataString(mod, b"kernel")])
```
Comment (pitrou) [inline llvmlite/tests/test_ir.py]: If `MetaDataString` starts accepting both bytes and unicode, we should probably test both here.

Comment (whitequark) [inline llvmlite/tests/test_ir.py]: It doesn't, that's why it's changed. It only accepts bytes and bytearray.

File: llvmlite/ir/values.py (hunk 1 above)
Comment (whitequark) [inline llvmlite/ir/values.py]: I did that (well, something similar).

File: llvmlite/ir/instructions.py (hunk 3 above)
Comment (pitrou) [inline llvmlite/ir/instructions.py]: Hmm, I see. It seems there are two conflicting constraints here:
- `_escape_string`, given its current docstring ("Escape the given bytestring for safe use as a LLVM array constant"), should not accept unicode strings
- metadata strings definitely feel like "text" and it would be awkward to refuse Python 3 string literals for them (Numba has a couple examples of them and I'm sure other users would, too)

So perhaps `_escape_string` should be renamed to e.g. `_escape_bytes` (for use by array constants), and a wrapper function `_escape_string` created (for use by metadata strings). What do you think?

Comment (whitequark) [inline llvmlite/ir/instructions.py]: No, this is completely wrong. What it "feels" is irrelevant. What matters is exactly one thing: how we consider `.ll` files to be encoded. There are two (reasonable) options:
 1. `.ll` files are ASCII, and UTF-8 cannot be put into them.
 2. `.ll` files are UTF-8-encoded.
`master` has a random mixture of these two. This patchset currently fixes it to represent option 1. If that's undesirable then it should be fixed to use option 2 (which I prefer anyway, since the instruction names are human-readable and LLVM doesn't care what's in there, and putting a `unicode` into a character array has a perfectly valid use, namely emitting UTF-8 string literals in a frontend).

Comment (whitequark) [inline llvmlite/ir/instructions.py]: Correction: the previous comment should say "the strings in the [abstract] LLVM IR", not "`.ll` files".

Comment (pitrou) [inline llvmlite/ir/instructions.py]: Well, the LLVM Language Reference doesn't say anything about the character encoding in which the IR is to be interpreted, so we have to find some compromise.

Arrays contain binary data, this is certain. Accepting text for them is at odds with the Python 3 model of separating text and bytes: if a frontend knows a string must be utf-8 encoded, therefore, it should encode it itself (after all, the string could be utf-16 encoded instead, e.g. for use with the Windows Unicode filesystem APIs).

I would lean towards considering metadata strings are human-readable text, but I suppose there may be cases where they are not. My concern here is that most existing code probably passes a `str` value (i.e. unicode on Python 3), not a `bytes`.

Comment (whitequark) [inline llvmlite/ir/instructions.py]: There is no real difference between allowing to pass `unicode` to metadata strings and to arrays. The latter is a way to communicate with your compiled program, the former is a way to communicate with various framework around your program. In one case you are assuming that your program should see human-readable text as UTF-8, in another you are assuming that your e.g. debugger or symbolizer should see human-readable text as UTF-8.

Metadata strings contain such things as function names and paths, neither of which are *required* to be UTF-8, or human-readable text for that matter. Indeed, on Linux, and depending on how you write your code, on Windows, there's no guarantee that paths are valid UTF-8.

And I think in both cases it's a reasonable assumption that any `unicode` passed is a human-readable text that should be converted to UTF-8, because if you're not representing human-readable text as UTF-8 then something is wrong with you.

Comment (pitrou) [inline llvmlite/ir/instructions.py]: > Metadata strings contain such things as function names and paths, neither of which are required to be UTF-8, or human-readable text for that matter.

That's true. However, those (function names, file paths) are precisely considered unicode strings on Python 3, so it makes sense that MetaDataString respects this. Or, at least, MetaDataString could encode them using the ASCII codec, which is safer (people would have to encode non-ASCII text manually, which prevents llvmlite from having to guess).

Comment (whitequark) [inline llvmlite/ir/instructions.py]: If so then we can simply restore the old commit but with a `.encode('ascii')` call rather than `.encode('utf-8')`. That would preserve the option 1 being the default behavior but also allow putting strings in various places.

**Comment by sklam** [comment]
To make it easier for users and to not breaking most of the existing code, I would prefer @pitrou's suggestion of:

>  Or, at least, MetaDataString could encode them using the ASCII codec, which is safer (people would have to encode non-ASCII text manually, which prevents llvmlite from having to guess).

**Comment by pitrou** [comment]
As a side note, the tests currently fail on 2.7:
https://travis-ci.org/numba/llvmlite/jobs/138080937#L624-L640

**Comment by whitequark** [comment]
So, any decision on this?

**Comment by pitrou** [comment]
I would selfishly defend the proposal that @sklam supported above :-) Basically, "practicality beats purity".

**Comment by whitequark** [comment]
Ok, that actually doesn't sound too bad to me.

**Comment by whitequark** [comment]
Updated, py2 issues fixed.

**Comment by pitrou** [comment]
Great, thanks!
